File: cpp/12_LevelOrderTraversal.cpp
/*
 * Q12. Binary Tree Level Order Traversal (Category: Trees)
 *
 * Problem: Return the level-order (BFS) traversal of a binary tree.
 *
 * Approach: BFS with a queue. Process each level by iterating queue.size() times.
 *
 * Time Complexity : O(n)
 * Space Complexity: O(n)
 */
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

vector<vector<int>> levelOrder(TreeNode* root) {
    vector<vector<int>> result;
    if (!root) return result;

    queue<TreeNode*> q;
    q.push(root);

    while (!q.empty()) {
        int levelSize = q.size();
        vector<int> level;
        for (int i = 0; i < levelSize; i++) {
            TreeNode* node = q.front(); q.pop();
            level.push_back(node->val);
            if (node->left)  q.push(node->left);
            if (node->right) q.push(node->right);
        }
        result.push_back(level);
    }
    return result;
}

int main() {
    //    3
    //   / \
    //  9  20
    //    /  \
    //   15    7
    TreeNode* root = new TreeNode(3);
    root->left  = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left  = new TreeNode(15);
    root->right->right = new TreeNode(7);

    auto res = levelOrder(root);
    for (auto& level : res) {
        cout << "[";
        for (int i = 0; i < (int)level.size(); i++) {
            cout << level[i];
            if (i + 1 < (int)level.size()) cout << ", ";
        }
        cout << "] ";
    }
    cout << endl; // [3] [9, 20] [15, 7]
    return 0;
}

File: python/12_level_order_traversal.py
"""
Q12. Binary Tree Level Order Traversal (Category: Trees)

Problem: Return the level-order (BFS) traversal of a binary tree.

Approach: BFS with a deque. Process each level by iterating len(queue) times,
          then add children for the next level.

Time Complexity : O(n)
Space Complexity: O(n)
"""
from __future__ import annotations
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val: int = 0,
                 left: Optional["TreeNode"] = None,
                 right: Optional["TreeNode"] = None):
        self.val = val
        self.left = left
        self.right = right


def level_order(root: Optional[TreeNode]) -> list[list[int]]:
    if root is None:
        return []
    result: list[list[int]] = []
    queue: deque[TreeNode] = deque([root])

    while queue:
        level = []
        for _ in range(len(queue)):
            node = queue.popleft()
            level.append(node.val)
            if node.left:  queue.append(node.left)
            if node.right: queue.append(node.right)
        result.append(level)

    return result


if __name__ == "__main__":
    #    3
    #   / \
    #  9  20
    #    /  \
    #   15    7
    root = TreeNode(3, TreeNode(9),
                    TreeNode(20, TreeNode(15), TreeNode(7)))
    print(level_order(root))  # [[3], [9, 20], [15, 7]]
